functionary/ui/forms/forms.py

```diff
@@ -31,20 +66,28 @@ def __init__(self, function, data=None):
         super().__init__(data)
 
         for param, value in function.schema["properties"].items():
-            req: bool = not value["default"]
+            req = not value["default"]
             initial = value.get("default", None)
+            param_type = get_param_type(value)
             kwargs = {
                 "label": value["title"],
-                "label_suffix": value["type"],
+                "label_suffix": param_type,
                 "initial": initial,
                 "required": req,
                 "help_text": value.get("description", None),
             }
-            field_class = _field_mapping.get(value["type"], None)
+
+            field_class, widget = _field_mapping.get(param_type, (None, None))
 
             if not field_class:
-                raise ValueError(f"Unknown field type for {param}: {value['type']}")
+                raise ValueError(f"Unknown field type for {param}: {param_type}")
+
+            if widget:
+                kwargs["widget"] = widget
+            if param_type in _transform_intitial_mapping:
+                kwargs["initial"] = _transform_intitial_mapping[param_type](initial)
```

Comment:
This call throws an exception if initial is `None`, so add that as part of your check before reaching this line.

Also, this might just be a me thing, but I keep reading `_transform_initial_mapping` as an action, which then makes my brain want to think that that's a function.  Which, maybe it could be?  What if instead you did:

```python
if widget:
    kwargs["widget"] = widget

kwargs["initial"] = _prepare_initial_value(param_type, initial)
```

And then the `_prepare_initial_value` call has all of the conditional logic, etc.  I know it's all kind of the same in the end.  Making it a function also has the added benefit of giving you a natural place for a simple docstring to explain what it's all about :-).

Reply:
Fixed